---
title: "isoform_analysis"
author: "daniellembecker"
date: "2024-04-15"
output: html_document
---

# Use processed isoform output files from RSEM output run on HPC outlined in this workflow: https://github.com/daniellembecker/DanielleBecker_Lab_Notebook/blob/master/_posts/2023-08-31-Acropora-pulchra-denovo-transcriptome.md
# isoform_expr.isoform.TPM.not_cross_norm file contains Transcripts Per Million (TPM) values, which are length-normalized expression values
# gene_to_isoform.map file is the mapping file that links gene identifiers to their corresponding isoform identifiers and positions

# Load required libraries
```{r}
library(ggplot2)
library(dplyr)
library(data.table)
```

# Load data frames
```{r}
# Load isoform expression data
isoform_expr <- read.delim("../../../isoform_expr.isoform.TPM.not_cross_norm", row.names = 1)
# "TRINITY_DN805066" - Prefix for de novo assembled transcript
# "c0" - Component or cluster number
# "g1" - Gene number within the component
# "i1" - Isoform identifier, denoting the first isoform of this gene

# Load extracted isoform positions
extracted.pos <- read.delim("../../../output_gene_to_isoform.map", header = FALSE)

#rename coloum 2 and 3 to positions
colnames(extracted.pos) <- c("isoform", "start", "stop")

# Load gene-to-isoform mapping
isoform_positions <- read.delim("../../../gene_to_isoform.map", header = FALSE)
colnames(isoform_positions) <- c("gene", "isoform")

#combine with isoform positiins dataframe
merged.isoforms <- left_join(isoform_positions, extracted.pos)

```

# Find out how many genes have more than one transcript mapping to them in a de novo transcriptome assembly,
```{r}
# Assuming your data is in a data frame called 'transcripts' with columns 'gene' and 'transcript'
genes_with_multiple_isoforms <- isoform_positions %>%
  group_by(gene) %>%
  filter(n() > 1) %>%
  summarise(num_isoforms = n()) %>%
  nrow()

print(paste0("Number of genes with more than one transcript: ", genes_with_multiple_isoforms))

#189,307 out of 1,476,390 = 12.8% of genes have more than one transcript
```


# Create plots to look at isoforms positions and expression for a given gene
```{r}
# Group data by gene
grouped_data <- merged.isoforms %>%
  group_by(gene)

# Create a new column with isoform ranges
grouped_data <- grouped_data %>%
  mutate(isoform_range = paste0(isoform, ": ", start, "-", stop))

# Subset the data to a smaller number of genes
subset_genes <- unique(grouped_data$gene)[1:20]  # Change the number to adjust the subset size
subset_data <- grouped_data[grouped_data$gene %in% subset_genes, ]

# Visualize isoform stacking for the subset of genes
ggplot(subset_data, aes(x = gene, y = isoform_range, fill = isoform)) +
  geom_tile() +
  scale_y_discrete(expand = c(0, 0)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))


```

# analyze the percentage of isoforms that are stacking (overlapping) in the same positions and same gene in your dataset
```{r}
# Function to check for overlapping isoforms
check_overlap <- function(start1, stop1, start2, stop2) {
  return(start1 <= stop2 && stop1 >= start2)
}

# Calculate stacking percentage for each gene
gene_stacking_percentages <- grouped_data %>%
  mutate(stacking = FALSE) %>%
  group_map(~ {
    gene_isoforms <- .x
    total_isoforms <- nrow(gene_isoforms)
    stacking_count <- 0
    
    for (i in 1:nrow(gene_isoforms)) {
      isoform1 <- gene_isoforms[i, ]
      start1 <- isoform1$start
      stop1 <- isoform1$stop
      
      for (j in 1:nrow(gene_isoforms)) {
        if (i != j) {
          isoform2 <- gene_isoforms[j, ]
          start2 <- isoform2$start
          stop2 <- isoform2$stop
          
          if (check_overlap(start1, stop1, start2, stop2)) {
            stacking_count <- stacking_count + 1
            gene_isoforms$stacking[i] <- TRUE
            gene_isoforms$stacking[j] <- TRUE
            break
          }
        }
      }
    }
    
    stacking_percentage <- (stacking_count / total_isoforms) * 100
    return(list(gene_isoforms = gene_isoforms, stacking_percentage = stacking_percentage))
  }) %>%
  transpose()

# Extract gene-wise stacking percentages
gene_stacking_percentages <- gene_stacking_percentages$stacking_percentage

# Calculate overall stacking percentage
overall_stacking_percentage <- sum(gene_stacking_percentages * sapply(gene_stacking_percentages, length)) / sum(sapply(gene_stacking_percentages, length))

# Print overall stacking percentage
print(paste0("Overall Percentage of Stacking Isoforms: ", round(overall_stacking_percentage, 2), "%"))

```










