---
title: "genetic differentiation for RNAseq SNPs  in heatwave adults and offspring"
author: "daniellembecker"
date: "20240924"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Use genetic differentiation outputs from PLINK computed on HPC from URI Andromeda to compare genetic differentiation among samples. 

## Install and load Libraries
```{r message = FALSE, warning = FALSE}
#install.packages(c("SNPRelate", "gdsfmt", "hierfstat"))
library(SNPRelate)
library(gdsfmt)
library(hierfstat)
```

## Read in genetic differentiation .vcf file made from original sample .bam files, called for SNP variants, and filtered, convert to genomic data structure object
```{r}
# Convert filtered .vcf file to GDS format
vcf.fn <- "bioinformatics/concatenated_output_sorted.vcf"  # Replace with your VCF file path
gds.fn <- "bioinformatics/output/heatwave_adults.gds"
snpgdsVCF2GDS(vcf.fn, gds.fn, method="biallelic.only")

# Open gds
genofile <- snpgdsOpen(gds.fn)

sample.id <- read.gdsn(index.gdsn(genofile, "sample.id"))
snp.id <- read.gdsn(index.gdsn(genofile, "snp.id"))
genotypes <- snpgdsGetGeno(genofile)

```

