---
title: "RNAseq Differential Gene Expression Analysis"
author: "Ariana S Huffmyer"
edited: "Danielle M Becker"
date: "05/20/2024"
output: html_document
editor_options: 
  chunk_output_type: console
---
Analysis of RNAseq Differential Gene Expression (DEG).  

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE) #Set Strings to character
```

# Set up 
Load required libraries.  

```{r}
if ("tidyverse" %in% rownames(installed.packages()) == 'FALSE') install.packages('tidyverse') 
if ("genefilter" %in% rownames(installed.packages()) == 'FALSE') install.packages('genefilter') 
if ("DESeq2" %in% rownames(installed.packages()) == 'FALSE') install.packages('DESeq2') 
if ("RColorBrewer" %in% rownames(installed.packages()) == 'FALSE') install.packages('RColorBrewer') 
if ("WGCNA" %in% rownames(installed.packages()) == 'FALSE') install.packages('WGCNA') 
if ("flashClust" %in% rownames(installed.packages()) == 'FALSE') install.packages('flashClust') 
if ("gridExtra" %in% rownames(installed.packages()) == 'FALSE') install.packages('gridExtra') 
if ("ComplexHeatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('ComplexHeatmap') 
if ("goseq" %in% rownames(installed.packages()) == 'FALSE') install.packages('goseq') 
if ("dplyr" %in% rownames(installed.packages()) == 'FALSE') install.packages('dplyr') 
if ("clusterProfiler" %in% rownames(installed.packages()) == 'FALSE') install.packages('clusterProfiler') 
if ("pheatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('pheatmap') 
if ("magrittr" %in% rownames(installed.packages()) == 'FALSE') install.packages('magrittr') 
if ("vegan" %in% rownames(installed.packages()) == 'FALSE') install.packages('vegan') 
if ("EnhancedVolcano" %in% rownames(installed.packages()) == 'FALSE') install.packages('EnhancedVolcano') 

#BiocManager::install('apeglm')
BiocManager::install('EnhancedVolcano')

library("tidyverse")
library("genefilter")
library("DESeq2")
library("RColorBrewer")
library("flashClust")
library("gridExtra")
library("goseq")
library("vegan")
library("factoextra")
library("dplyr")
library("clusterProfiler")
library("pheatmap")
library("magrittr")
library("vegan")
library("apeglm")
library("EnhancedVolcano")
library("cowplot")
library("viridis")
library("pheatmap")
```

# Data input and filtering  

Import the data files.  

```{r}
#load metadata sheet with sample name and developmental stage information
metadata <- read.csv("bioinformatics/data/heatwave_CGA_extractions_metadata.csv", header = TRUE, sep = ",")%>%select(bioinf.ID, treatment, type, parent)
metadata$code<-paste0(metadata$treatment, "_", metadata$parent)
head(metadata)

#load gene count matrix generated from cluster computation
gcount <- as.data.frame(read.csv("bioinformatics/data/gene_count_apul_matrix.csv", row.names="gene_id"), colClasses = double)
head(gcount)
```

Check that there are no genes with 0 counts across all samples. 

```{r}
dim(gcount) 

gcount<-gcount %>%
     mutate(Total = rowSums(.[, 1:12]))%>%
    filter(!Total==0)%>%
    select(!Total)

dim(gcount)
```

We started with 160,088 transcripts, which is the total number in the annotation. About 89,849 genes that had 0's across all samples (not detected) with 70,239 remaining.

Conduct data filtering, this includes:  

*pOverA*: Specifying the minimum count for a proportion of samples for each gene. Here, we are using a pOverA of 0.05. This is because we have 12 samples with a minimum of n=3 samples  per group. Therefore, we will accept genes that are present in 3/12 = 0.25 of the samples because we expect different expression by treatment group (parent x temperature). We are further setting the minimum count of genes to 10, such that 25% of the samples must have a gene count of >10 in order for the gene to remain in the data set.  

Filter in the package "genefilter". Pre-filtering our dataset to reduce the memory size dataframe, increase the speed of the transformation and testing functions, and improve quality of statistical analysis by removing low-coverage counts. Removed counts could represent outliers in the data and removing these improves sensitivity of statistical tests.

```{r}
filt <- filterfun(pOverA(0.25,10))

#create filter for the counts data
gfilt <- genefilter(gcount, filt)

#identify genes to keep by count filter
gkeep <- gcount[gfilt,]

#identify gene lists
gn.keep <- rownames(gkeep)

#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keep),])

#How many rows do we have before and after filtering?
nrow(gcount) #Before
nrow(gcount_filt) #After
```

Before filtering, we had approximately 70,239 genes. After filtering for pOverA, we have approximately 21,510 genes. This indicates that there were 21,510 genes present in 25% of samples at <10 counts per gene.  

Prepare sample names. Keep only sample ID number in the colnames of gcount_filt.  
```{r}
# Function to extract numbers after 'X' and remove everything after '_sorted'
extract_sample_numbers <- function(col_names) {
  gsub("^X(\\d+)_sorted.*$", "\\1", col_names)
}

# Apply the function to all column names
new_col_names <- sapply(names(gcount_filt), extract_sample_numbers)

# Rename the columns
names(gcount_filt) <- new_col_names

# Output the modified dataframe column names
names(gcount_filt)
length(names(gcount_filt))
```

Column names are now correct and there are 12 as expected.  

Print the bioif.ID names in the metadata file. 
```{r}
metadata$bioinf.ID
```

In order for the DESeq2 algorithms to work, the SampleIDs on the metadata file and count matrices have to match exactly and in the same order. They are.  

# Format metadata and count data columns  

Convert treatment metadata categories to factors. 
```{r}
metadata$treatment<-as.factor(metadata$treatment)
metadata$parent<-as.factor(metadata$parent)
metadata$type<-as.factor(metadata$type)
metadata$code<-as.factor(metadata$code)
```

Set levels of factors. 
```{r}
metadata$treatment<-factor(metadata$treatment, levels=c("ambient", "heatwave"))
metadata$parent<-factor(metadata$parent, levels=c("ambient", "heatwave"))
metadata$type<-factor(metadata$type, levels=c("larvae", "adult"))
```

Make sure the metadata and the columns in the gene count matrix are all the same.  
```{r}
metadata$bioinf.ID
colnames(gcount_filt)
```

Reorder them automatically if needed.  
```{r}
list<-colnames(gcount_filt)
list<-as.factor(list)

metadata$bioinf.ID<-as.factor(metadata$bioinf.ID)

# Re-order the levels
metadata$bioinf.ID <- factor(as.character(metadata$bioinf.ID), levels=list)
# Re-order the data.frame
metadata_ordered <- metadata[order(metadata$bioinf.ID),]
metadata_ordered$bioinf.ID

head(metadata_ordered)

metadata_ordered$bioinf.ID
colnames(gcount_filt)
head(gcount_filt)
```

These match we are good to continue to the next step!    

# Construct DESeq2 data set first for adults between treatments

Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at treatments in adules to test for any differences in gene expression across heatwave treatment. 

The model we will use will analyze the effect of heatwave treatment on gene expression in the adults. We do not have random or batch effects in this study to analyze.  

Filter dataframes for just adults between heatwave treatments

```{r}
# Assuming your metadata dataframe is called 'metadata'
adult_metadata <- metadata %>%
  filter(type == "adult")  # Adjust 'type' to match your column name

# Make sure the row names of the metadata match the column names of the count matrix
rownames(adult_metadata) <- adult_metadata$bioinf.ID  # Adjust 'sample_id' to match your column name

# Assuming your gene count matrix is called 'gcount_filt'
adult_counts <- gcount_filt[, rownames(adult_metadata)]

adult_metadata$bioinf.ID
colnames(adult_counts)
```

```{r}
#Set DESeq2 design
gdds_adults <- DESeqDataSetFromMatrix(countData = adult_counts,
                              colData = adult_metadata,
                             design = ~treatment)
```

First we are going to log-transform the data using a variance stabilizing transforamtion (VST). This is only for visualization purposes. Essentially, this is roughly similar to putting the data on the log2 scale. It will deal with the sampling variability of low counts by calculating within-group variability (if blind=FALSE). Importantly, it does not use the design to remove variation in the data, and so can be used to examine if there may be any variability do to technical factors such as extraction batch effects.

This function calculates a variance stabilizing transformation (VST) from the fitted dispersion-mean relation(s) and then transforms the count data (normalized by division by the size factors or normalization factors), yielding a matrix of values which are now approximately homoskedastic (having constant variance along the range of mean values).    

To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4.  

To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4.

Chunk should return TRUE if <4.  
```{r}
SF.gdds_adults <- estimateSizeFactors(gdds_adults) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than 4 for us to use vst
print(sizeFactors(SF.gdds_adults)) #View size factors

all(sizeFactors(SF.gdds_adults)) < 4
```

All size factors are less than 4, so we can use VST transformation.  

```{r}
gvst_adults <- vst(gdds_adults, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst_adults), 3) #view transformed gene count data for the first three genes in the dataset. 

dim(gvst_adults)
```

# Plot a heatmap of sample to sample distances and all genes and of heatmap between treatments

```{r}
gsampleDists_adults <- dist(t(assay(gvst_adults))) #calculate distance matix
gsampleDistMatrix_adults <- as.matrix(gsampleDists_adults) #distance matrix
rownames(gsampleDistMatrix_adults) <- colnames(gvst_adults) #assign row names
colnames(gsampleDistMatrix_adults) <- NULL #assign col names
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #assign colors

save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
   stopifnot(!missing(x))
   stopifnot(!missing(filename))
   pdf(filename, width=width, height=height)
   grid::grid.newpage()
   grid::grid.draw(x$gtable)
   dev.off()
}

pht <- pheatmap(gsampleDistMatrix_adults, #plot matrix
         clustering_distance_rows = gsampleDists_adults, #cluster rows
         clustering_distance_cols = gsampleDists_adults, #cluster columns
         col = colors) #set colors

save_pheatmap_pdf(pht, "heatwave/output/sample_distances.pdf")
dev.off()


```

# Conduct PERMANOVA and PCA for all genes

Export data for PERMANOVA test.  
```{r}
test<-t(assay(gvst_adults)) #export as matrix
test<-as.data.frame(test)

#add category columns
test$bioinf.ID<-rownames(test)
test$treatment <- adult_metadata$treatment[match(test$bioinf.ID, adult_metadata$bioinf.ID)]
```

Build PERMANOVA model for all genes in the dataset.  
```{r}
dim(test)
scaled_test <-prcomp(test[c(1:21510)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test)

# scale data
vegan <- scale(test[c(1:21510)])

# PerMANOVA 
permanova<-adonis2(vegan ~ treatment, data = test, method='eu')
permanova
```
 

## Examine PCA and sample distances of all genes  

Plot a PCA of samples by treatment between adults for all genes.        
```{r}
gPCAdata_adults <- plotPCA(gvst_adults, intgroup = c("treatment"), returnData=TRUE)
percentVar_adults <- round(100*attr(gPCAdata_adults, "percentVar")) #plot PCA of samples with all data

adults_PCA <- ggplot(gPCAdata_adults, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) + 
  ggtitle(expression(bold("All genes (21,510 genes)")))+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +  
  scale_color_manual(name="Treatment", values=c("ambient"="blue", "heatwave"="brown4"))+
  theme_classic() + 
  theme(panel.border = element_blank(),
        axis.line = element_line(colour = "black"),
        plot.background=element_blank());adults_PCA

ggsave("heatwave/output/adult_pca_allgenes.png", adults_PCA, width=8, height=4)
```

Add a plot with elipses.  

```{r}
adults_PCA_ellipse <- adults_PCA + stat_ellipse(); adults_PCA_ellipse

ggsave("figures/rna_seq/pca_ellipse.pdf", allgenes_ellipse, width=9, height=5)
ggsave("figures/rna_seq/pca_ellipse.jpeg", allgenes_ellipse, width=9, height=5)
```

# Run DEG analysis of adult heatwave treatments

Use the likelihood ratio approach based on guidance from DESeq2: "DESeq2 offers two kinds of hypothesis tests: the Wald test, where we use the estimated standard error of a log2 fold change to test if it is equal to zero, and the likelihood ratio test (LRT). The LRT examines two models for the counts, a full model with a certain number of terms and a reduced model, in which some of the terms of the full model are removed. The test determines if the increased likelihood of the data using the extra terms in the full model is more than expected if those extra terms are truly zero.

The LRT is therefore useful for testing multiple terms at once, for example testing 3 or more levels of a factor at once, or all interactions between two variables. The LRT for count data is conceptually similar to an analysis of variance (ANOVA) calculation in linear regression, except that in the case of the Negative Binomial GLM, we use an analysis of deviance (ANODEV), where the deviance captures the difference in likelihood between a full and a reduced model." 

## 1. View DEG's by treatment

Run Wald test.  
```{r, message = FALSE}
DEG_T <- DESeq(gdds_adults, test="Wald")
```

View results by treatment effects. Create dataframe for each comparison and add a column for the specific contrast.  
```{r}
resultsNames(DEG_T)
#this shows the comparisons made (shown similar to lm output)

adult_deg<-results(DEG_T, contrast=c("treatment","heatwave","ambient"))
adult_deg <- as.data.frame(subset(adult_deg, padj<0.01))
adult_deg <- as.data.frame(subset(adult_deg, abs(log2FoldChange)>1.5))
adult_deg$contrast <- c("adult treatment")
adult_deg$gene <- rownames(adult_deg)
rownames(adult_deg) <- NULL

dim(adult_deg)

```

There are 57 genes differentially expressed between adult heatwave treatmments. In these contrasts, positive fold changes indicate higher expression in the first listed contrast.  

```{r}
dim(adult_deg)
length(unique(adult_deg$gene))
```
There are 57 total genes and 57 unique genes.

Get a VST normalized gene count matrix for these DEGs. 
```{r}
DEG_adult_vst <- gdds_adults[unique(adult_deg$gene)]
dim(DEG_adult_vst)

DEG_adult_vst <- varianceStabilizingTransformation(DEG_adult_vst) 
```

### Plot a volcano plot for adult heatwave comparison.   

Adults between heatwave and ambient treatment
```{r}
res <- results(DEG_T, contrast=c("treatment","heatwave","ambient"))

volcano_plot_adults<-EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = 'Adult Treatments',
    pCutoff = 0.01,
    FCcutoff = 1.5,
    pointSize = 3.0,
    labSize = 0.0, 
    legendPosition = 'none');volcano_plot_adults
```

Save volcano plot. 

```{r}
ggsave("heatwave/output/volcano_adult_treatment.png", plot=volcano_plot_adults, width=6, height=6)
```

### Plot a heatmap for genes that are differentially expressed between treatments.

Set themes and metadata.  
```{r}
# Make a matrix for computing similarity
mat <- assay(DEG_adult_vst) # make an expression object
mat <- mat - rowMeans(mat) #difference in expression compared to average across all samples
dim(mat)
ann_colors <- list(treatment= c(ambient="blue", heatwave="brown4"))
df_DEGSeq2_adults <- as.data.frame(colData(DEG_adult_vst)[c("treatment")]) #make dataframe for column naming and associated treatment

# Sort the columns of the matrix by treatment
treatment_order <- order(df_DEGSeq2_adults$treatment)
mat <- mat[, treatment_order]
df_DEGSeq2_adults <- df_DEGSeq2_adults[treatment_order, , drop = FALSE]


DEGSeq2_heatmap_adults <- pheatmap(mat, scale= "row", legend=TRUE, annotation_legend=TRUE, annotation_col=df_DEGSeq2_adults, annotation_colors = ann_colors,
                            clustering_distance_rows="euclidean", clustering_method = "average",
                            show_rownames =FALSE,
                            show_colnames =TRUE,
                            cluster_cols = FALSE)

DEGSeq2_heatmap_adults
```

Save heatmap of adult DEGs plot. 

```{r}
ggsave("heatwave/output/DEG_heatmap_adults.png", plot=DEGSeq2_heatmap_adults, width=8, height=5)
```


### Run a PERMANOVA and plot a PCA for treatment comparisons.  

Conduct PERMANOVA by parent and temperature 

Export data for PERMANOVA test.  
```{r}
test_DEG_temp<-t(assay(DEG_adult_vst)) #export as matrix
test_DEG_temp<-as.data.frame(test_DEG_temp)

#add category columns
test_DEG_temp$bioinf.ID <-rownames(test_DEG_temp)
test_DEG_temp$temperature<-metadata$temperature[match(test_DEG_temp$sample, metadata$sample)]
test_DEG_temp$parent<-metadata$parent[match(test_DEG_temp$sample, metadata$sample)]
```

Build PERMANOVA model for DEG's that are different between temperatures.  
```{r}
dim(test_DEG_temp)

scaled_test_DEG_temp <-prcomp(test_DEG_temp[c(1:936)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test_DEG_temp)

# scale data
vegan_DEG_temp <- scale(test_DEG_temp[c(1:936)])

# PerMANOVA 
permanova_DEG<-adonis2(vegan_DEG_temp ~ parent*temperature, data = test_DEG_temp, method='eu')
permanova_DEG
```
 

Plot a PCA of DEG's for temperature.        
```{r}
gPCAdata_DEG_adult <- plotPCA(DEG_adult_vst, intgroup = c("treatment"), returnData=TRUE)
percentVar_DEG_temp <- round(100*attr(gPCAdata_DEG_adult, "percentVar")) #plot PCA of samples with all data

DEG_PCA_adult <- ggplot(gPCAdata_DEG_adult, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) + 
  ggtitle(expression(bold("DEGs: adult treatment (57 genes)")))+
  xlab(paste0("PC1: ",percentVar_DEG_temp[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_DEG_temp[2],"% variance")) +
  scale_color_manual(name="Treatment", values=c("ambient"="blue", "heatwave"="brown4"))+
#  geom_text(x=0, y=-9.5, label="p[Temperature]=0.001", color="black")+
  theme_classic() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines 
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); DEG_PCA_adult

ggsave("heatwave/output/pca_DEG_adults.png", DEG_PCA_adult, width=5, height=4)
```

From this we can see very clear separation by treatment with high variance explained by PC1. The difference in parent exists in these DEG's, explained by PC2 with lower variance explained.

## 2. View DEG's by parents to larvae

###########still need to edit##########

```{r, message = FALSE}
#DEG_P <- DESeq(gdds, test="LRT", full=~temperature+parent+temperature:parent, reduced=~temperature)
DEG_P <- DESeq(gdds, test="Wald")
```

View results by parental effects. Create dataframe for each comparison and add a column for the specific contrast.  
```{r}
resultsNames(DEG_P)
#this shows the comparisons made (shown similar to lm output in terms of interactions)
#The interaction terms temperature33.parentBleached and temperature33.parentNonbleached give the difference between the parental effect for a given temperature and the parental effect for the reference parent (Wildtype)

BvsWT_deg<-results(DEG_P, contrast=c("parent","Bleached","Wildtype"))
BvsWT_deg <- as.data.frame(subset(BvsWT_deg, padj<0.01))
BvsWT_deg <- as.data.frame(subset(BvsWT_deg, abs(log2FoldChange)>1.5))
BvsWT_deg$contrast <- c("BvsWT")
BvsWT_deg$gene <- rownames(BvsWT_deg)
rownames(BvsWT_deg) <- NULL

dim(BvsWT_deg)

```
There are 207 genes differentially expressed between bleached and wildtype. There are 112 genes differentially expressed between nonbleached and wildtype. There are 282 genes differentially expressed between nonbleached and bleached. In these contrasts, positive fold changes indicate higher expression in the first listed contrast.  

Combine this list of DEGs. 

```{r}
dim(DEG_parent)
length(unique(DEG_parent$gene))
```
There are 601 total genes and 432 unique genes, indicating that some are shared in the contrasts. 

Get a VST normalized gene count matrix for these DEGs. 
```{r}
DEG_parent_vst <- gdds[unique(DEG_parent$gene)]
dim(DEG_parent_vst)

DEG_parent_vst <- varianceStabilizingTransformation(DEG_parent_vst) 
```

### Plot a volcano plot for these comparisons.   

B vs WT
```{r}
res <- results(DEG_P, contrast=c("parent","Bleached","Wildtype"))

volcano_plot_BvsWT<-EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = 'Bleached vs Wildtype',
    pCutoff = 0.01,
    FCcutoff = 1.5,
    pointSize = 3.0,
    labSize = 0.0, 
    legendPosition = 'none');volcano_plot_BvsWT
```

NB vs WT
```{r}
res <- results(DEG_P, contrast=c("parent","Nonbleached","Wildtype"))

volcano_plot_NBvsWT<-EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = 'Nonbleached vs Wildtype',
    pCutoff = 0.01,
    FCcutoff = 1.5,
    pointSize = 3.0,
    labSize = 0.0, 
    legendPosition = 'none');volcano_plot_NBvsWT
```

NB vs B 
```{r}
res <- results(DEG_P, contrast=c("parent","Nonbleached","Bleached"))

volcano_plot_NBvsB<-EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = 'Nonbleached vs Bleached',
    pCutoff = 0.01,
    FCcutoff = 1.5,
    pointSize = 3.0,
    labSize = 0.0, 
    legendPosition = 'none');volcano_plot_NBvsB
```

Plot all volcano plots together. 

```{r}
volcano_parent_grid<-plot_grid(volcano_plot_BvsWT, volcano_plot_NBvsWT, volcano_plot_NBvsB, align="vh", nrow=1, ncol=3)

ggsave("figures/rna_seq/volcano_parent.pdf", plot=volcano_parent_grid, width=18, height=6)
ggsave("figures/rna_seq/volcano_parent.png", plot=volcano_parent_grid, width=18, height=6)
```

### Plot a heatmap for genes that are differentially expressed between parent phenotypes 

Set themes and metadata.  
```{r}
legend_names_col = colnames(assay(DEG_parent_vst))

df <- as.data.frame(colData(gdds)[,c("temperature","parent")])
df$temperature <- factor(as.character(df$temperature), levels=c("27", "30", "33"))
df$parent <- factor(as.character(df$parent), levels=c("Wildtype", "Bleached", "Nonbleached"))
df<-df%>%
  arrange(temperature, parent)

ann_colors = list(
    temperature = c("27" = "lightblue", "30" = "pink2","33"= "red2"), 
    parent = c("Bleached" = "orange", "Nonbleached" = "brown4", "Wildtype" = "gray")
)
```

Plot heatmap.  
```{r}
pdf("figures/rna_seq/DEG_heatmap_parent.pdf", width=15, height=20)
pheatmap(assay(DEG_parent_vst), cluster_rows=TRUE, show_rownames=FALSE, color=inferno(10), show_colnames=TRUE, fontsize_row=3, scale="row", cluster_cols=TRUE, annotation_col=df, annotation_colors = ann_colors, labels_col=legend_names_col)
dev.off()
```

### Run a PERMANOVA and plot a PCA for parent comparisons.  

Conduct PERMANOVA by parent and temperature of parental DEGs. 

Export data for PERMANOVA test.  
```{r}
test_DEG_parent<-t(assay(DEG_parent_vst)) #export as matrix
test_DEG_parent<-as.data.frame(test_DEG_parent)

#add category columns
test_DEG_parent$sample<-rownames(test_DEG_parent)
test_DEG_parent$temperature<-metadata$temperature[match(test_DEG_parent$sample, metadata$sample)]
test_DEG_parent$parent<-metadata$parent[match(test_DEG_parent$sample, metadata$sample)]
```

Build PERMANOVA model for DEG's that are different between temperatures.  
```{r}
dim(test_DEG_parent)

scaled_test_DEG_parent <-prcomp(test_DEG_parent[c(1:432)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test_DEG_parent)

# scale data
vegan_DEG_parent <- scale(test_DEG_parent[c(1:432)])

# PerMANOVA 
permanova_DEG_parent<-adonis2(vegan_DEG_parent ~ parent*temperature, data = test_DEG_parent, method='eu')
permanova_DEG_parent
```
About 35% of variance is explained by first PC followed by 15% on the second PC and <10% on the third PC. 

Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

adonis2(formula = vegan_DEG_parent ~ parent * temperature, data = test_DEG_parent, method = "eu")
                   Df SumOfSqs      R2       F Pr(>F)    
parent              2  11344.8 0.49549 35.8707  0.001 ***
temperature         2   3117.4 0.13615  9.8567  0.001 ***
parent:temperature  4   1317.8 0.05756  2.0834  0.010 ** 
Residual           45   7116.0 0.31080                   
Total              53  22896.0 1.00000                

Parent is significant as expected because we are analyzing parental DEGs. There are still temperature effects, which also makes sense because there was such a strong temperature effect on global expression. Interestingly, there is a parent x temperature interaction, indicating response to temperature is mediated by parental phenotype. We will look into this more next.   

Plot a PCA of DEG's for parent        
```{r}
gPCAdata_DEG_parent <- plotPCA(DEG_parent_vst, intgroup = c("temperature", "parent"), returnData=TRUE)
percentVar_DEG_parent <- round(100*attr(gPCAdata_DEG_parent, "percentVar")) #plot PCA of samples with all data

DEG_PCA_parent <- ggplot(gPCAdata_DEG_parent, aes(PC1, PC2, color=parent, shape=temperature)) + 
  geom_point(size=3) + 
  ggtitle(expression(bold("DEGs: parent (432 genes)")))+
  xlab(paste0("PC1: ",percentVar_DEG_parent[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_DEG_parent[2],"% variance")) +
  scale_color_manual(name="Parental Phenotype", values=c("Wildtype"="gray", "Bleached"="orange", "Nonbleached"="brown4"))+
  scale_shape_manual(name="Temperature", values=c(19,17,15))+
  geom_text(x=0, y=10, label="p[Parent]=0.001", color="black")+
  geom_text(x=0, y=8, label="p[Temperature]=0.001", color="black")+
  geom_text(x=0, y=6, label="p[Interaction]=0.010", color="black")+
  theme_classic() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines 
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); DEG_PCA_parent

ggsave("figures/rna_seq/pca_DEG_parent.pdf", DEG_PCA_parent, width=8, height=4)
ggsave("figures/rna_seq/pca_DEG_parent.png", DEG_PCA_parent, width=8, height=4)
```

From this we can see very clear separation by parent with high variance explained by PC1 separating parent. There is significant separation between temperatures within parent with 33°C being especially different, as expected from temperature DEGs. Interestingly, gene expression of wildtype parents at 27-30°C is separated from the other parental phenotypes.  

## 3. View DEG's differential between parents at elevated temperatures.  

Run differential expression test using LRT testing for significant genes by parent by removing the effect of temperature:parent and comparing the significance of genes in the full model (`parent + temperature + parent:temperature`) to the reduced model removing the interactive effect.   
```{r, message = FALSE}
#DEG_I <- DESeq(gdds, test="LRT", full=~temperature+parent+temperature:parent, reduced=~temperature+parent)
DEG_I <- DESeq(gdds, test="Wald")
```

Next, view genes that are uniquely differentially expressed between parental types at 30°C and 33°C. In other words, view genes that are different between NB and B groups at 30°C and 33°. Conduct this for 30°C and then 33°C and then join the lists.   

To do this, I will use contrast DEGs between `temperature30.parentNonbleached`  and `temperature30.parentBleached` or `temperature33.parentNonbleached`  and `temperature33.parentBleached` in the DESeq2 results. This will examine the differential effect of parental treatment at 30°C and 33°C. Each of the NB and B parental groups have been compared to the wildtype group as a reference, so, this will contrast the comparisons of wildtype-NB and wildtype-B at the 30°C/33°C treatment. This will provide us with DEG's between parent groups at temperature. This will produce a list of genes that contribute to interactive effects.   

I am changing the p-value to 0.05 and the fold change to >1 for this test. This is because we have far fewer genes detected in the interaction.  

Create dataframe for each comparison and add a column for the specific contrast.  
```{r}
resultsNames(DEG_I)
#this shows the comparisons made (shown similar to lm output in terms of interactions)

NBvsB_30_deg<-results(DEG_I, contrast=list("temperature30.parentNonbleached", "temperature30.parentBleached"))
NBvsB_30_deg <- as.data.frame(subset(NBvsB_30_deg, padj<0.05))
NBvsB_30_deg <- as.data.frame(subset(NBvsB_30_deg, abs(log2FoldChange)>1))
NBvsB_30_deg$contrast <- c("NBvsB_30C")
NBvsB_30_deg$gene <- rownames(NBvsB_30_deg)
rownames(NBvsB_30_deg) <- NULL

dim(NBvsB_30_deg)

NBvsB_33_deg<-results(DEG_I, contrast=list("temperature33.parentNonbleached", "temperature33.parentBleached"))
NBvsB_33_deg <- as.data.frame(subset(NBvsB_33_deg, padj<0.05))
NBvsB_33_deg <- as.data.frame(subset(NBvsB_33_deg, abs(log2FoldChange)>1))
NBvsB_33_deg$contrast <- c("NBvsB_33C")
NBvsB_33_deg$gene <- rownames(NBvsB_33_deg)
rownames(NBvsB_33_deg) <- NULL

dim(NBvsB_33_deg)
```

There are 7 DEGs between NB and B at 30°C. There are 49 at 33°C. In these contrasts, positive fold changes indicate higher expression in the first listed contrast - in this case, Nonbleached samples.  

Combine this list of DEGs into one dataframe. 

```{r}
DEG_interaction<-rbind(NBvsB_30_deg, NBvsB_33_deg)
dim(DEG_interaction)
length(unique(DEG_interaction$gene))
```
There are 56 total genes and 55 unique genes, indicating that only one is shared between 30°C and 33°C.   

```{r}
#DEG_interaction_unique <- anti_join(DEG_interaction, DEG_parent, by = "gene")

DEG_interaction_unique<-DEG_interaction

dim(DEG_interaction_unique)
str(DEG_interaction_unique)
length(unique(DEG_interaction_unique$gene))
```
There are 55 genes that are uniquely differentially expressed between NB and B at elevated temperature (30 or 33°C). 

Get a VST normalized gene count matrix for these DEGs. 
```{r}
DEG_interaction_vst <- gdds[unique(DEG_interaction_unique$gene)]
dim(DEG_interaction_vst)

DEG_interaction_vst <- varianceStabilizingTransformation(DEG_interaction_vst) 
```

### Upset plot of genes between temperature contrasts 

Create an upset diagram showing genes unique and shared in each temperature comparison.  

```{r}
list1_interaction<-DEG_interaction_unique%>%
  filter(contrast=="NBvsB_30C")
dim(list1_interaction)

list2_interaction<-DEG_interaction_unique%>%
  filter(contrast=="NBvsB_33C")
dim(list2_interaction)

ven_interaction <- venndetail(list("NBvsB_30C" = list1_interaction$gene, "NBvsB_33C" = list2_interaction$gene))

upset_interaction<-plot(ven_interaction, type = "upset");upset_interaction

pdf("figures/rna_seq/upset_interaction.pdf", width=8, height=6)
plot(upset_interaction, type = "upset")
dev.off()
```
There are 48 DEGs at 33°C and 6 at 30°C with 1 shared between the contrasts.   

### Plot a volcano plot for these comparisons.   

NB vs B at 30°C
```{r}
res <- results(DEG_I, contrast=list("temperature30.parentNonbleached", "temperature30.parentBleached"))

volcano_plot_NBvsB_30C<-EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = 'Nonbleached vs Bleached at 30°C',
    pCutoff = 0.05,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 0.0, 
    legendPosition = 'none');volcano_plot_NBvsB_30C
```

NB vs B at 33°C
```{r}
res <- results(DEG_I, contrast=list("temperature33.parentNonbleached", "temperature33.parentBleached"))

volcano_plot_NBvsB_33C<-EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue',
    title = 'Nonbleached vs Bleached at 33°C',
    pCutoff = 0.05,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 0.0, 
    legendPosition = 'none');volcano_plot_NBvsB_33C
```

Plot all volcano plots together. 

```{r}
volcano_interaction_grid<-plot_grid(volcano_plot_NBvsB_30C, volcano_plot_NBvsB_33C, align="vh", nrow=1, ncol=2)

ggsave("figures/rna_seq/volcano_interaction.pdf", plot=volcano_interaction_grid, width=12, height=6)
ggsave("figures/rna_seq/volcano_interaction.png", plot=volcano_interaction_grid, width=12, height=6)
```

### Plot a heatmap for genes that are differentially expressed between parent phenotypes at elevated temperature 

Set themes and metadata.  
```{r}
legend_names_col = colnames(assay(DEG_interaction_vst))

df <- as.data.frame(colData(gdds)[,c("temperature","parent")])
df$temperature <- factor(as.character(df$temperature), levels=c("27", "30", "33"))
df$parent <- factor(as.character(df$parent), levels=c("Wildtype", "Bleached", "Nonbleached"))
df<-df%>%
  arrange(temperature, parent)

ann_colors = list(
    temperature = c("27" = "lightblue", "30" = "pink2","33"= "red2"), 
    parent = c("Bleached" = "orange", "Nonbleached" = "brown4", "Wildtype" = "gray")
)
```

Plot heatmap.  
```{r}
pdf("figures/rna_seq/DEG_heatmap_interaction.pdf", width=9, height=5)
pheatmap(assay(DEG_interaction_vst), cluster_rows=TRUE, show_rownames=FALSE, color=inferno(10), show_colnames=TRUE, fontsize_row=3, scale="row", cluster_cols=TRUE, annotation_col=df, annotation_colors = ann_colors, labels_col=legend_names_col)
dev.off()
```

### Run a PERMANOVA and plot a PCA for interaction comparisons

Conduct PERMANOVA by parent and temperature of interaction DEGs. 

Export data for PERMANOVA test.  
```{r}
test_DEG_interaction<-t(assay(DEG_interaction_vst)) #export as matrix
test_DEG_interaction<-as.data.frame(test_DEG_interaction)

#add category columns
test_DEG_interaction$sample<-rownames(test_DEG_interaction)
test_DEG_interaction$temperature<-metadata$temperature[match(test_DEG_interaction$sample, metadata$sample)]
test_DEG_interaction$parent<-metadata$parent[match(test_DEG_interaction$sample, metadata$sample)]
```

Build PERMANOVA model for DEG's that are different between NB and B at 30°C 
```{r}
dim(test_DEG_interaction)

scaled_test_DEG_interaction <-prcomp(test_DEG_interaction[c(1:55)], scale=TRUE, center=TRUE)
fviz_eig(scaled_test_DEG_interaction)

# scale data
vegan_DEG_interaction <- scale(test_DEG_interaction[c(1:55)])

# PerMANOVA 
permanova_DEG_interaction<-adonis2(vegan_DEG_interaction ~ parent*temperature, data = test_DEG_interaction, method='eu')
permanova_DEG_interaction
```
About 30% of variance is explained by first PC followed by 15% on the second PC and 10% on the third PC. 

Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

adonis2(formula = vegan_DEG_interaction ~ parent * temperature, data = test_DEG_interaction, method = "eu")
                   Df SumOfSqs      R2       F Pr(>F)    
parent              2   586.68 0.20126 11.5799  0.001 ***
temperature         2   731.82 0.25105 14.4448  0.001 ***
parent:temperature  4   456.57 0.15663  4.5059  0.001 ***
Residual           45  1139.93 0.39106                   
Total              53  2915.00 1.00000                   

Interaction is significant as expected. Strong parent and temperature effects are also present as expected.   

Plot a PCA of DEG's for the interaction

```{r}
gPCAdata_DEG_interaction <- plotPCA(DEG_interaction_vst, intgroup = c("temperature", "parent"), returnData=TRUE)
percentVar_DEG_interaction <- round(100*attr(gPCAdata_DEG_interaction, "percentVar")) #plot PCA of samples with all data

DEG_PCA_interaction <- ggplot(gPCAdata_DEG_interaction, aes(PC1, PC2, color=parent, shape=temperature)) + 
  geom_point(size=3) + 
  ggtitle(expression(bold("DEGs: NB vs B at 30-33°C (55 genes)")))+
  xlab(paste0("PC1: ",percentVar_DEG_parent[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_DEG_parent[2],"% variance")) +
  scale_color_manual(name="Parental Phenotype", values=c("Wildtype"="gray", "Bleached"="orange", "Nonbleached"="brown4"))+
  scale_shape_manual(name="Temperature", values=c(19,17,15))+
  geom_text(x=1, y=-2.5, label="p[Parent]=0.001", color="black")+
  geom_text(x=1, y=-3, label="p[Temperature]=0.001", color="black")+
  geom_text(x=1, y=-3.5, label="p[Interaction]=0.001", color="black")+
  theme_classic() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines 
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); DEG_PCA_interaction

ggsave("figures/rna_seq/pca_DEG_interaction.pdf", DEG_PCA_interaction, width=8, height=4)
ggsave("figures/rna_seq/pca_DEG_interaction.png", DEG_PCA_interaction, width=8, height=4)
```

From this we can see very clear separation by parent at 33°C with little separation and higher variability in other groups/temperatures. This is what we expected to see with DEGs that are different at elevated temperatures between parent types. 

# Save lists of gene counts and DEG information 

## Temperature

DEG_temperature: List of DEG's and fold change/pvalues 
DEG_temperature_vst: Count matrix of genes from DEG list 

```{r}
deg_temperature_list<-DEG_temperature%>%
  select(gene, everything())
write_csv(deg_temperature_list, "output/rna_seq/temperature_deg_list.csv")

dim(DEG_temperature_vst)
deg_temperature_counts<-t(assay(DEG_temperature_vst))
deg_temperature_counts<-as.data.frame(deg_temperature_counts)
dim(deg_temperature_counts)
deg_temperature_counts$sample<-rownames(deg_temperature_counts)
deg_temperature_counts<-deg_temperature_counts%>%
  select(sample, everything())
write_csv(deg_temperature_counts, "output/rna_seq/temperature_deg_counts.csv")
```

## Parent

DEG_parent: List of DEG's and fold change/pvalues 
DEG_parent_vst: Count matrix of genes from DEG list 

```{r}
deg_parent_list<-DEG_parent%>%
  select(gene, everything())
write_csv(deg_parent_list, "output/rna_seq/parent_deg_list.csv")

dim(DEG_parent_vst)
deg_parent_counts<-t(assay(DEG_parent_vst))
deg_parent_counts<-as.data.frame(deg_parent_counts)
dim(deg_parent_counts)
deg_parent_counts$sample<-rownames(deg_parent_counts)
deg_parent_counts<-deg_parent_counts%>%
  select(sample, everything())
write_csv(deg_parent_counts, "output/rna_seq/parent_deg_counts.csv")
```

## NB vs B at elevated temperature

DEG_interaction_unique: List of DEG's and fold change/pvalues
DEG_interaction_vst: Count matrix of genes from DEG list 

```{r}
dim(DEG_interaction_unique)
deg_interaction_list<-DEG_interaction_unique%>%
  select(gene, everything())
write_csv(deg_interaction_list, "output/rna_seq/NBvsB_30-33_deg_list.csv")

dim(DEG_interaction_vst)
deg_interaction_counts<-t(assay(DEG_interaction_vst))
deg_interaction_counts<-as.data.frame(deg_interaction_counts)
dim(deg_interaction_counts)
deg_interaction_counts$sample<-rownames(deg_interaction_counts)
deg_interaction_counts<-deg_interaction_counts%>%
  select(sample, everything())
write_csv(deg_interaction_counts, "output/rna_seq/NBvsB_30-33_deg_counts.csv")
```

## All genes detected 

List of all genes in the dataset (gvst) = 24005 genes which will serve as a reference for functional enrichment.   

```{r}
dim(gvst)
all_genes<-t(assay(gvst))
all_genes<-as.data.frame(all_genes)
dim(all_genes)
all_genes$sample<-rownames(all_genes)
all_genes<-all_genes%>%
  select(sample, everything())

write_csv(file="output/rna_seq/all_genes.csv", all_genes)
```

